### PR TITLE
[rtefsutils, rtemodel, cbuild] Fix local file URLs handling (#221)

### PR DESCRIPTION
**Pull request by grasci-arm**
Fixes issue https://github.com/Open-CMSIS-Pack/devtools/issues/299

### CODE REVIEW

**Review by brondani — APPROVED** [review]

**Review by ErikLundinSTM — COMMENTED** [review]
`GetAbsPathFromLocalUrl()` accepts URLs with an extra leading slash in the path, compensating for the incorrect test cases (see separate comments) which in violence with RFC 8089 have one extra leading slash.

In addition, Windows paths end up with a leading slash, e.g. `file://localhost/c:/WINDOWS/clock.avi` --> `/c:/WINDOWS/clock.avi` and `file:///c:/WINDOWS/clock.avi` --> `/c:/WINDOWS/clock.avi` (examples from [Wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme)).

File: libs/rtefsutils/src/RteFsUtils.cpp [hunk 1]
```diff
@@ -755,4 +755,26 @@ bool RteFsUtils::FindFileRegEx(const vector<string>& searchPaths, const string&
   return false;
 }
 
+string RteFsUtils::GetAbsPathFromLocalUrl(const string& url) {
+  string filepath = url;
+  // File URI scheme allows the literal 'localhost' to be omitted entirely
+  // or may contain an empty hostname
+  static const string&& fileScheme = "file:/";
```
[the remark this answers is not in the document]
Comment (brondani) [reply libs/rtefsutils/src/RteFsUtils.cpp:762]: @ErikLundinSTM From the [document](https://en.wikipedia.org/wiki/File_URI_scheme) that you mentioned the pattern `file:/` is valid, the other patterns are complementary to it.

> A valid file URI must therefore begin with either file:/path (no hostname), file:///path (empty hostname), or file://hostname/path

File: libs/rtefsutils/test/src/RteFsUtilsTest.cpp [hunk 2]
```diff
@@ -1069,3 +1069,25 @@ TEST_F(RteFsUtilsTest, FindFileRegEx_NoMatch) {
   EXPECT_EQ(false, RteFsUtils::FindFileRegEx(searchPaths, ".*\\.cdefault\\.yml", finding));
   RteFsUtils::RemoveDir(testdir);
 }
+
+TEST_F(RteFsUtilsTest, GetAbsPathFromLocalUrl) {
+  const string& absoluteFilenameRegular = RteFsUtils::AbsolutePath(filenameRegular).generic_string();
+  // Local host
+  const string& testUrlLocalHost = "file://localhost/" + absoluteFilenameRegular;
```
[the remark this answers is not in the document]
Comment (brondani) [reply libs/rtefsutils/test/src/RteFsUtilsTest.cpp:1076]: Test cases that you consider correct/incorrect in Linux have the opposite behaviour in Windows.
I guess we should separate the test cases according to the platform.

File: libs/rtefsutils/src/RteFsUtils.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ErikLundinSTM) [reply libs/rtefsutils/src/RteFsUtils.cpp:762]: I get the point. However, conceptually the first two slashes belong together (even if they are optional if not providing any host), like this: `file:[//[host]]/path/to/file`. The end result in this case is the same. This was just a small nitpick.

**Comment by brondani** [comment]
> In addition, Windows paths end up with a leading slash, e.g. `file://localhost/c:/WINDOWS/clock.avi` --> `/c:/WINDOWS/clock.avi` and `file:///c:/WINDOWS/clock.avi` --> `/c:/WINDOWS/clock.avi` (examples from [Wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme)).

@ErikLundinSTM When running this [test case](https://github.com/Open-CMSIS-Pack/devtools/blob/main/libs/rtefsutils/test/src/RteFsUtilsTest.cpp#L1077) in Windows, I see: `file://localhost/C:/RteFsUtilsTest/dir/subdir/file.txt` --> `C:/RteFsUtilsTest/dir/subdir/file.txt`

File: libs/rtefsutils/test/src/RteFsUtilsTest.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ErikLundinSTM) [reply libs/rtefsutils/test/src/RteFsUtilsTest.cpp:1076]: Yes, covering both Linux and Windows paths is certainly the way to go. I would suggest not depending on `AbsolutePath()` in the test (no need to use the CWD of the file system where the test happens to be run) and instead hard code some paths. This also has the advantage of avoiding file I/O in the unit test.

**Comment by ErikLundinSTM** [comment]
I see. I got the extra slash since I ran it on Linux, so depending on the OS, `GetAbsPathFromLocalUrl()` may return different paths for the same URL. So if you run the test from somewhere below `C:\`, does it use that root, or does it use the one from the URL? Example: Will `file://localhost/d:/somepath` return `d:/somepath` or something based on your CWD?

**Comment by brondani** [comment]
> Example: Will `file://localhost/d:/somepath` return `d:/somepath` or something based on your CWD?

It returns what is in the URL, `d:/somepath` in your example.
The root path prepending only takes place when running on Unix.
Maybe it could be better abstracted in the crossplatform, I will take a look.

**Comment by brondani** [comment]
@ErikLundinSTM Thanks for your observations, they were addressed in this PR: 
https://github.com/Open-CMSIS-Pack/devtools/pull/320

**Comment by ErikLundinSTM** [comment]
> @ErikLundinSTM Thanks for your observations, they were addressed in this PR: #320

Great!

**Pull request merged**